File: src/runtime/python/grammar.py
#! coding=UTF-8

# Copyright 2012 Thomas Bollmeier <tbollmeier@web.de>

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#   http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# ===== Interne Objekte: =====

class Connectable(object):

    def __init__(self):
        pass

    def getSocket(self):
        raise NotImplementedError

class Socket(object):

    def __init__(self):
        pass

    def getSuccessors(self, context):
        raise NotImplementedError

class Pluggable(object):

    def __init__(self):
        pass

    def getPlug(self):
        raise NotImplementedError

class Plug(object):

    def __init__(self):
        pass

    def connectTo(self, socket):
        raise NotImplementedError

class GrammarElement(object):

    def __init__(self):
        pass

    def connect(self, successorElement):
        raise NotImplementedError

# ===== API =====

class Rule(Connectable, Pluggable, GrammarElement):

    def __init__(self, name='', identifier=''):

        if self.__class__ == Rule:
            raise Exception('Rule is an abstract class')

        if not name:
            name = self.__class__.__name__

        Connectable.__init__(self)
        Pluggable.__init__(self)
        GrammarElement.__init__(self)

        self._start = RuleStartNode(self, name, identifier)
        self._end = RuleEndNode(self)
        self._envVars = {}

    def expand(self, start, end, context):

        raise NotImplementedError

    # Knoten in AST ggf. umwandeln:
    def transform(self, astNode):

        return astNode

    def getName(self):

        return self._start.getName()

    def onSuccRequested(self, start, end, context):

        self.expand(start, end, context)

    def getEndNode(self):

        return self._end

    def getEnvVars(self):

        return self._envVars

    def setEnvVar(self, name, value=True):

        self._envVars[name] = value

    def getSocket(self):

        return self._start

    def getPlug(self):

        return self._end

    def connect(self, successorElement):

        self._end.connectTo(successorElement.getSocket())

        return successorElement

def defineRule(name):

    return _RuleFactory(name)

class initialize(object):

    def __init__(self, ruleFactory):

        self._ruleFactory = ruleFactory

    def __call__(self, initFunc):

        self._ruleFactory._initFunc = initFunc

class expand(object):

    def __init__(self, ruleFactory):

        self._ruleFactory = ruleFactory

    def __call__(self, expandFunc):

        self._ruleFactory._expandFunc = expandFunc

class transform(object):

    def __init__(self, ruleFactory):

        self._ruleFactory = ruleFactory

    def __call__(self, transformFunc):

        self._ruleFactory._transformFunc = transformFunc

class Grammar(Rule):

    def __init__(self, tokenTypes):

        Rule.__init__(self)

        self._tokenTypes = tokenTypes

    def getTokenTypes(self):

        return self._tokenTypes

def tokenNode(tokenType, identifier=''):

    if not tokenType:
        raise Exception('Token type not specified')

    return TokenNode(tokenType, identifier)

def connector():

    return PlugNode(Node.TECHNICAL)

def connect(predecessor, successor):

    predecessor.getPlug().connectTo(successor.getSocket())

    return successor

def zeroToOne(element):

    return Multiplier(element, Multiplier.ZERO_TO_ONE)

def zeroToMany(element):

    return Multiplier(element, Multiplier.ZERO_TO_MANY)

def oneToMany(element):

    return Multiplier(element, Multiplier.ONE_TO_MANY)

def sequence(*elements):

    return Sequence(elements)

def fork(*branches):

    return Fork(branches)

class Switch(Connectable, Pluggable, GrammarElement):

    def __init__(self, branches):

        Connectable.__init__(self)
        Pluggable.__init__(self)
        GrammarElement.__init__(self)

        self._end = connector()
        self._start = _SwitchNode(branches, self._end)

    def getSocket(self):

        return self._start

    def getPlug(self):

        return self._end

    def connect(self, successorElement):

        self._end.connectTo(successorElement.getSocket())

        return successorElement

class Condition(Connectable, Pluggable, GrammarElement):

    def __init__(self, conditionFunc):

        Connectable.__init__(self)
        Pluggable.__init__(self)
        GrammarElement.__init__(self)

        self._end = connector()
        self._start = _ConditionalNode(conditionFunc, self._end)

    def getSocket(self):

        return self._start

    def getPlug(self):

        return self._end

    def connect(self, successorElement):

        self._end.connectTo(successorElement.getSocket())

        return successorElement

class SuccessorError(Exception):

    pass

# ===== Interne Objekte: =====

class _CustomRule(Rule):

    def __init__(self,
                 name,
                 ident,
                 initFunc,
                 expandFunc,
                 transformFunc
                 ):

        Rule.__init__(self, name, ident)

        if initFunc:
            initFunc(self)

        self._expandFunc = expandFunc
        self._transformFunc = transformFunc

    def expand(self, start, end, context):

        self._expandFunc(start, end, context)

    def transform(self, astNode):

        if self._transformFunc:
            return self._transformFunc(astNode)
        else:
            return astNode

class _RuleFactory(object):

    def __init__(self, name):

        self._name = name
        self._initFunc = None
        self._expandFunc = None
        self._transformFunc = None

    def __call__(self, ident=''):

        return _CustomRule(self._name,
                           ident,
                           self._initFunc,
                           self._expandFunc,
                           self._transformFunc
                           )

    def getName(self):

        return self._name

    name = property(getName)

class IdNode(object):

    def __init__(self):
        pass

    def getId(self):
        raise NotImplementedError

class Node(Connectable, Socket):

    # Node categories:
    RULE_START = 1
    RULE_END = 2
    TOKEN = 3
    TECHNICAL = 4

    __nextTechId = 1

    def __init__(self, category, tokenType = None):

        Connectable.__init__(self)
        Socket.__init__(self)

        self._catg = category
        self._tokenType = tokenType

        self.__techId = Node.__nextTechId
        Node.__nextTechId += 1

    def getTokenType(self):

        return self._tokenType

    def getTokenTypeId(self):

        return self._tokenType and self._tokenType.getId() or -1

    def isRuleStart(self):

        return self._catg == Node.RULE_START

    def isRuleEnd(self):

        return self._catg == Node.RULE_END

    def isTokenNode(self):

        return self._catg == Node.TOKEN

    def isTechnicalNode(self):

        return self.catg == Node.TECHNICAL

    def getSocket(self):

        return self

    def getTechnicalId(self):

        return self.__techId

class PlugNode(Node, Pluggable, Plug, GrammarElement):

    def __init__(self, category, tokenType = None):

        Node.__init__(self, category, tokenType)
        Pluggable.__init__(self)
        Plug.__init__(self)
        GrammarElement.__init__(self)

        self._successors = []

    def getSuccessors(self, context):

        return self._successors

    def getPlug(self):

        return self

    def connectTo(self, socket):

        if not socket in self._successors:
            self._successors.append(socket)

    def connect(self, successorElement):

        self.connectTo(successorElement.getSocket())

        return successorElement

class RuleStartNode(Node, IdNode):

    def __init__(self, ruleAccess, name, identifier):

        Node.__init__(self, Node.RULE_START)
        IdNode.__init__(self)

        self._ruleAccess = ruleAccess
        self._name = name
        self._id = identifier

    def getSuccessors(self, context):

        start = PlugNode(Node.TECHNICAL)
        end = PlugNode(Node.TECHNICAL)

        self._ruleAccess.onSuccRequested(start, end, context)

        end.connectTo(self._ruleAccess.getEndNode())

        return [start]

    def getEnvVars(self):

        return self._ruleAccess.getEnvVars()

    def getName(self):

        return self._name

    def getId(self):

        return self._id

class RuleEndNode(PlugNode):

    def __init__(self, ruleAccess):

        PlugNode.__init__(self, Node.RULE_END, None)

        self._ruleAccess = ruleAccess

    def transform(self, astNode):

        return self._ruleAccess.transform(astNode)

class TokenNode(PlugNode, IdNode):

    def __init__(self, tokenType, identifier = ''):

        PlugNode.__init__(self, Node.TOKEN, tokenType)
        IdNode.__init__(self)

        self._id = identifier
        self._envVarChangeFunc = None
        self._envVarUndoFunc = None

    def getId(self):

        return self._id

    def setEnvChange(self, changeFunc, undoFunc):

        self._envVarChangeFunc = changeFunc
        self._envVarUndoFunc = undoFunc

    def changesEnv(self):

        return ( self._envVarChangeFunc is not None ) and \
          ( self._envVarUndoFunc is not None )

    def changeEnv(self, envVars, token):

        if self._envVarChangeFunc:
            self._envVarChangeFunc(envVars, token, self)

    def undoEnvChange(self, envVars, token):

        if self._envVarUndoFunc:
            self._envVarUndoFunc(envVars, token, self)

class _SwitchNode(Node):

    def __init__(self, branches, end):

        Node.__init__(self, Node.TECHNICAL)

        self._branches = branches
        self._end = end

    def getSuccessors(self, context):

        keyword = context.getCurKeyword()
        if not keyword:
            return []

        try:
            branch = self._branches[keyword]
            start = connector()
            start.connect(branch).connect(self._end)
            return [start]
        except KeyError:
            return []

class _ConditionalNode(Node):

    def __init__(self, conditionFunc, end):

        Node.__init__(self, Node.TECHNICAL)

        self._conditionFunc = conditionFunc
        self._end = end

    def getSuccessors(self, context):

        if self._conditionFunc(context):
            return [self._end]
        else:
            raise SuccessorError

class RuleInternalAccess(object):

    def __init__(self):
        pass

    def getName(self):
        raise NotImplementedError

    def getEndNode(self):
        raise NotImplementedError

    def getEnvVars(self):
        raise NotImplementedError

    def onSuccRequested(self, start, end, context):
        raise NotImplementedError

    def transform(self, astNode):
        raise NotImplementedError

class Multiplier(Connectable, Pluggable, GrammarElement):

    ZERO_TO_ONE = 1
    ZERO_TO_MANY = 2
    ONE_TO_MANY = 3

    def __init__(self, element, mult):

        Connectable.__init__(self)
        Pluggable.__init__(self)
        GrammarElement.__init__(self)

        self._start = connector()
        self._end = connector()

        if mult == Multiplier.ZERO_TO_ONE:

            connect(self._start, self._end)
            connect(self._start, element)
            connect(element, self._end)

        elif mult == Multiplier.ZERO_TO_MANY:

            connect(self._start, self._end)
            connect(self._start, element)
            connect(element, self._start)

        elif mult == Multiplier.ONE_TO_MANY:

            connect(self._start, element)
            connect(element, self._end)
            connect(element, element)

    def getSocket(self):

        return self._start

    def getPlug(self):

        return self._end

    def connect(self, successorElement):

        self._end.connectTo(successorElement.getSocket())

        return successorElement

class Sequence(Connectable, Pluggable, GrammarElement):

    def __init__(self, elements):

        Connectable.__init__(self)
        Pluggable.__init__(self)
        GrammarElement.__init__(self)

        self._start = connector()
        self._end = connector()

        current = self._start
        for elem in elements:
            connect(current, elem)
            current = elem
        connect(current, self._end)

    def getSocket(self):

        return self._start

    def getPlug(self):

        return self._end

    def connect(self, successorElement):

        self._end.connectTo(successorElement.getSocket())

        return successorElement

class Fork(Connectable, Pluggable, GrammarElement):

    def __init__(self, branches):

        Connectable.__init__(self)
        Pluggable.__init__(self)
        GrammarElement.__init__(self)

        self._start = connector()
        self._end = connector()

        for branch in branches:
            self._start.connect(branch).connect(self._end)

    def getSocket(self):

        return self._start

    def getPlug(self):

        return self._end

    def connect(self, successorElement):

        self._end.connectTo(successorElement.getSocket())

        return successorElement

File: src/runtime/php/grammar.php
<?php

/*
 * Copyright 2012 Thomas Bollmeier <tbollmeier@web.de>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

abstract class Bovinus_Rule implements Bovinus_Connectable, Bovinus_Pluggable, Bovinus_GrammarElement, Bovinus_RuleInternalAccess {

    public function __construct($name, $identifier = "") {

        $this->_start = new Bovinus_RuleStartNode($this, $name, $identifier);
        $this->_end = new Bovinus_RuleEndNode($this);
        $this->_envVars = array();
    }

    public function getName() {

        return $this->_start->getName();
    }

    public function getEndNode() {

        return $this->_end;
    }

    public function getEnvVars() {

        return $this->_envVars;
    }

    public function setEnvVar($name, $value = TRUE) {

        $this->_envVars[$name] = $value;
    }

    public function onSuccRequested($start, $end, $context) {

        $this->expand($start, $end, $context);
    }

    abstract public function expand($start, $end, $context);

    public function transform($astNode) {

        return $astNode;
    }

    public function connect(Bovinus_Connectable $successorElement) {

        $this->_end->connectTo($successorElement->getSocket());

        return $successorElement;
    }

    public function getPlug() {

        return $this->_end;
    }

    public function getSocket() {

        return $this->_start;
    }

    private $_start;
    private $_end;
    private $_envVars;

}

abstract class Bovinus_Grammar extends Bovinus_Rule {

    public function __construct($name, $tokenTypes) {

        parent::__construct($name);

        $this->tokenTypes = $tokenTypes;
    }

    public function getTokenTypes() {

        return $this->tokenTypes;
    }

    private $tokenTypes;

}

function bovinus_tokenNode($tokenType, $identifier = "") {

    return new Bovinus_TokenNode($tokenType, $identifier);
}

function bovinus_sequence() {

    $elements = func_get_args();

    return new Bovinus_Sequence($elements);
}

function bovinus_fork() {

    $branches = func_get_args();

    return new Bovinus_Fork($branches);
}

function bovinus_zero_to_one($element) {
    
    return new Bovinus_Multiplier($element, Bovinus_Multiplier::ZERO_TO_ONE);
}

function bovinus_zero_to_many($element) {
    
    return new Bovinus_Multiplier($element, Bovinus_Multiplier::ZERO_TO_MANY);
}

function bovinus_one_to_many($element) {
    
    return new Bovinus_Multiplier($element, Bovinus_Multiplier::ONE_TO_MANY);
}

interface Bovinus_Connectable {

    function getSocket();
}

interface Bovinus_Socket {

    function getSuccessors($context);
}

interface Bovinus_Pluggable {

    function getPlug();
}

interface Bovinus_Plug {

    function connectTo($socket);
}

interface Bovinus_GrammarElement {

    function connect(Bovinus_Connectable $successorElement);
}

interface Bovinus_IdNode {

    function getId();
}

interface Bovinus_RuleInternalAccess {

    function getName();

    function getEndNode();

    function getEnvVars();

    function onSuccRequested($start, $end, $context);

    function transform($astNode);
}

abstract class Bovinus_Node implements Bovinus_Connectable, Bovinus_Socket {
// Node categories:

    const RULE_START = 1;
    const RULE_END = 2;
    const TOKEN = 3;
    const TECHNICAL = 4;

    public function __construct($category, $tokenType = null) {

        $this->catg = $category;
        $this->tokenType = $tokenType;
        $this->techId = Bovinus_Node::$nextTechId++;
    }

    public function getTokenType() {

        return $this->tokenType;
    }

    public function getTokenTypeId() {

        if ($this->tokenType != null) {
            return $this->tokenType->getId();
        } else {
            return -1;
        }
    }

    public function isRuleStart() {

        return $this->catg == Bovinus_Node::RULE_START;
    }

    public function isRuleEnd() {

        return $this->catg == Bovinus_Node::RULE_END;
    }

    public function isTokenNode() {

        return $this->catg == Bovinus_Node::TOKEN;
    }

    public function isTechnicalNode() {

        return $this->catg == Bovinus_Node::TECHNICAL;
    }

    public function getSocket() {

        return $this;
    }

    public function getTechnicalId() {

        return $this->techId;
    }

    static private $nextTechId = 1;
    private $catg;
    private $tokenType;
    private $techId;

}

class Bovinus_PlugNode extends Bovinus_Node implements Bovinus_Pluggable, Bovinus_Plug, Bovinus_GrammarElement {

    public static function createConnector() {

        return new Bovinus_PlugNode(Bovinus_Node::TECHNICAL);
    }

    public function __construct($category, $tokenType = null) {

        parent::__construct($category, $tokenType);

        $this->successors = array();
    }

    public function getSuccessors($context) {

        return $this->successors;
    }

    public function getPlug() {

        return $this;
    }

    public function connectTo($socket) {

        if (array_search($socket, $this->successors, TRUE) === FALSE) {
            array_push($this->successors, $socket);
        }
    }

    public function connect(Bovinus_Connectable $successorElement) {

        $this->connectTo($successorElement->getSocket());

        return $successorElement;
    }

    private $successors;

}

class Bovinus_RuleStartNode extends Bovinus_Node implements Bovinus_IdNode {

    public function __construct(Bovinus_RuleInternalAccess $ruleAccess, $name, $identifier) {

        parent::__construct(Bovinus_Node::RULE_START);

        $this->ruleAccess = $ruleAccess;
        $this->name = $name;
        $this->ident = $identifier;
    }

    public function getId() {

        return $this->ident;
    }

    public function getSuccessors($context) {

        $start = Bovinus_PlugNode::createConnector();
        $end = Bovinus_PlugNode::createConnector();

        $this->ruleAccess->onSuccRequested($start, $end, $context);

        $end->connectTo($this->ruleAccess->getEndNode());

        return array($start);
    }

    public function getEnvVars() {

        return $this->ruleAccess->getEnvVars();
    }

    public function getName() {

        return $this->name;
    }

    private $ruleAccess;
    private $name;
    private $ident;

}

class Bovinus_RuleEndNode extends Bovinus_PlugNode {

    public function __construct(Bovinus_RuleInternalAccess $ruleAccess) {

        parent::__construct(Bovinus_Node::RULE_END);

        $this->ruleAccess = $ruleAccess;
    }

    public function transform($astNode) {

        return $this->ruleAccess->transform($astNode);
    }

    private $ruleAccess;

}

interface Bovinus_EnvVarChanger {

    function change($envVars, $token, $node);

    function undo($envVars, $token, $node);
}

class Bovinus_TokenNode extends Bovinus_PlugNode implements Bovinus_IdNode {

    public function __construct($tokenType, $identifier = "") {

        parent::__construct(Bovinus_Node::TOKEN, $tokenType);

        $this->ident = $identifier;
    }

    public function getId() {

        return $this->ident;
    }

    public function setEnvChanger(Bovinus_EnvVarChanger $changer) {

        $this->envVarChanger = $changer;
    }

    public function changesEnvVars() {

        return ($this->envVarChanger != null);
    }

    public function changeEnv($envVars, $token) {

        if ($this->envVarChanger != null) {
            $this->envVarChanger->change($envVars, $token, $this);
        }
    }

    public function undoEnvChange($envVars, $token) {

        if ($this->envVarChanger != null) {
            $this->envVarChanger->undo($envVars, $token, $this);
        }
    }

    private $ident = "";
    private $envVarChanger = null;

}

class Bovinus_Sequence implements Bovinus_Connectable, Bovinus_Pluggable, Bovinus_GrammarElement {

    public function __construct($elements) {

        $this->start = Bovinus_PlugNode::createConnector();
        $this->end = Bovinus_PlugNode::createConnector();

        $prev = $this->start;

        foreach ($elements as $element) {
            $prev->connect($element);
            $prev = $element;
        }

        $prev->connect($this->end);
    }

    public function connect(Bovinus_Connectable $successorElement) {

        $this->end->connectTo($successorElement->getSocket());

        return $successorElement;
    }

    public function getPlug() {

        return $this->end;
    }

    public function getSocket() {

        return $this->start;
    }

    private $start = null;
    private $end = null;

}

class Bovinus_Fork implements Bovinus_Connectable, Bovinus_Pluggable, Bovinus_GrammarElement {

    public function __construct($branches) {

        $this->start = Bovinus_PlugNode::createConnector();
        $this->end = Bovinus_PlugNode::createConnector();

        foreach ($branches as $branch) {
            $this->start->connect($branch)->connect($this->end);
        }
    }

    public function connect(Bovinus_Connectable $successorElement) {

        $this->end->connectTo($successorElement->getSocket());

        return $successorElement;
    }

    public function getPlug() {

        return $this->end;
    }

    public function getSocket() {

        return $this->start;
    }

    private $start;
    private $end;

}

class Bovinus_Multiplier implements Bovinus_Connectable, Bovinus_Pluggable, Bovinus_GrammarElement {

    const ZERO_TO_ONE = 1;
    const ZERO_TO_MANY = 2;
    const ONE_TO_MANY = 3;

    public function __construct($element, $multiplicity) {

        $this->start = Bovinus_PlugNode::createConnector();
        $this->end = Bovinus_PlugNode::createConnector();

        switch ($multiplicity) {

            case Bovinus_Multiplier::ZERO_TO_ONE:
                $this->start->connect($element)->connect($this->end);
                $this->start->connect($this->end);
                break;

            case Bovinus_Multiplier::ZERO_TO_MANY:
                $this->start->connect($element)->connect($this->start);
                $this->start->connect($this->end);
                break;

            case Bovinus_Multiplier::ONE_TO_MANY:
                $this->start->connect($element);
                $element->connect($this->end);
                $element->connect($element);
                break;

            default:
                throw new Exception("Unknown multiplicity '" . $multiplicity . "'");
                break;
        }
    }

    public function connect(Bovinus_Connectable $successorElement) {

        $this->end->connectTo($successorElement->getSocket());

        return $successorElement;
    }

    public function getPlug() {

        return $this->end;
    }

    public function getSocket() {

        return $this->start;
    }

    private $start;
    private $end;

}

class Bovinus_Condition implements Bovinus_Connectable, Bovinus_Pluggable, Bovinus_GrammarElement {

    public function __construct($conditionFunc) {

        $this->start = null;
        $this->end = Bovinus_PlugNode::createConnector();
    }

    public function connect(Bovinus_Connectable $successorElement) {

        $this->end->connectTo($successorElement->getSocket());

        return $successorElement;
    }

    public function getPlug() {

        return $this->end;
    }

    public function getSocket() {

        return $this->start;
    }

    private $start;
    private $end;

}

class Bovinus_ConditionalNode extends Bovinus_Node {

    public function __construct(Bovinus_ConditionChecker $checker, $endNode) {

        parent::__construct(Bovinus_Node::TECHNICAL);

        $this->checker = $checker;
        $this->end = $endNode;
    }

    public function getSuccessors($context) {

        if ($this->checker->isFulfilled($context)) {
            return $this->end;
        } else {
            throw new Bovinus_SuccessorError();
        }
    }

    private $checker;
    private $end;

}

class Bovinus_SuccessorError extends Exception {

    public function __construct() {

        parent::__construct("Unexpected successor!");
    }

}

interface Bovinus_ConditionChecker {

    function isFulfilled($context);
}
